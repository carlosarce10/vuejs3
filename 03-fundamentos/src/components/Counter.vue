<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
  <p data-testid="counter">{{ counter }}</p>

  <div>
    <button @click="decrement()">-1</button>
    <button @click="increment()">+1</button>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    start: {
      type: Number,
      default: 100,
      //required: true,
      validator(value) {
        return value >= 0;
      },
    },
  },
  data() {
    return {
      counter: this.start,
      candyCount: 0,
      newtitle: '',
    };
  },
  methods: {
    getSquare() {
      return this.counter * this.counter;
    },
    decrement() {
      this.counter--;
    },
    increment() {
      this.counter++;
    },
  },
  computed: {
    squareCounter() {
      return this.counter * this.counter;
    },
    customTitle() {
      return this.title || 'Counter';
    },
  },
};
</script>

<style scoped>
button {
  background-color: rgb(15, 140, 223);
  color: white;
  border-radius: 5px;
  margin: 0, 5px;
  border: 1px solid white;
  cursor: pointer;
  padding: 5px, 15px;
  transition: 0.3s ease-in-out;
  width: 60px;
  height: 40px;
}
button:hover {
  background-color: rgb(15, 125, 199);
  transition: 0.3s ease-in-out;
}
</style>
