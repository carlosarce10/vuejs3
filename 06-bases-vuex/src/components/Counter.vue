<template>
  <h1>Counter - Vuex</h1>
  <h2>Direct access: {{ $store.state.count }}</h2>
  <h2>Computed: {{ countComputed }}</h2>
  <button @click="increment">+1</button>
  <button @click="incrementBy">+5</button>
  <button @click="randomInt" :disabled="isLoading">+random</button>

  <h1>mapState</h1>
  <h2>mapState: {{ count }}</h2>
  <h2>state renombrado: {{ counter }}</h2>

  <h2>Direct getter: {{ $store.getters.squareCount }}</h2>
</template>

<script>
import { mapState, mapActions } from 'vuex';

export default {
  data() {
    return {};
  },

  computed: {
    countComputed() {
      return this.$store.state.count;
    },
    ...mapState(['count', 'isLoading']),

    // renombrar variable del state
    ...mapState({
      counter: (state) => state.count,
    }),
  },
  methods: {
    increment() {
      this.$store.commit('increment');
    },
    incrementBy() {
      this.$store.commit('incrementBy', 5);
    },

    //...mapActions(['incrementRandom']),

    // renombrar action
    ...mapActions({
      randomInt: 'incrementRandom',
    }),
  },
};
</script>
