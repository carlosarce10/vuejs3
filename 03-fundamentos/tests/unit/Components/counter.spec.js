import { shallowMount } from '@vue/test-utils';
import Counter from '@/components/Counter';

describe('Counter component', () => {
  let wrapper;

  // -- Forma de reutilizar el wrapper
  beforeEach(() => {
    wrapper = shallowMount(Counter);
  });

  /* test('debe de hacer match con el snapshot', () => {
    const wrapper = shallowMount(Counter);
    expect(wrapper.html()).toMatchSnapshot();
  }); */

  test('h2 debe tener el valor por defecto "Counter"', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy();
    const h2 = wrapper.find('h2');
    expect(h2.text()).toBe('Counter');
  });

  test('el valor por defecto debe ser 100 en el p', () => {
    // const p = wrapper.findAll('p');
    // expect(p[1].text()).toBe('100');
    const value = wrapper.find('[data-testid="counter"]').text();
    expect(value).toBe('100');
  });

  test('Debe incrementar y decrementar el valor del contador', async () => {
    // -- DestructuraciÃ³n de la lista
    const [decreaseBtn, increaseBtn] = wrapper.findAll('button');

    await decreaseBtn.trigger('click');
    await increaseBtn.trigger('click');
    await increaseBtn.trigger('click');

    const value = wrapper.find('[data-testid="counter"]').text();
    expect(value).toBe('101');
  });

  // -- Testeando props
  test('debe de establecer el valor por defecto', async () => {
    const { start } = wrapper.props();
    const value = wrapper.find('[data-testid="counter"]').text();

    expect(Number(value)).toBe(start);
  });

  // -- Testeando envio de proposa
  test('debe mostrar la prop title', () => {
    const title = 'Hola mundo';
    // -- envio de props
    const wrapper = shallowMount(Counter, {
      props: {
        title,
      },
    });

    expect(wrapper.find('h2').text()).toBe(title);
  });
});
