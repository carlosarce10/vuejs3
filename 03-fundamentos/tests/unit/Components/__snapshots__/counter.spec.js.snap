// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component debe de hacer match con el snapshot 1`] = `
<h2>Counter</h2>
<p>5 <sup>2</sup> = 25</p>
<div><button>-1</button><button>+1</button></div>
`;
